matchmaker.js

```diff
@@ -0,0 +1,146 @@
+        if (joining_player === undefined) {
+            console.log("join_matchmaking Player is undefined")
+            return false
+        }
+
+        var player_join_version = json_data.version
+        var deck_id = json_data.deck_id
+        joining_player.set_deck_id(deck_id)
+        var desired_match_type = json_data.match_type
+        
+        if (this.banned_characters.includes(deck_id)) {
+            const message = {
+                type: 'matchmaking_failed',
+                reason: 'banned_matchmaking_character'
+            }
+            ws.send(JSON.stringify(message))
+            return true
+        }
+
+        var success = false
+
+        if (this.match_queues[desired_match_type]["has_player_waiting"] == false) {
+            match_queues[desired_match_type]["has_player_waiting"] = true
+            match_queues[desired_match_type]["waiting_player"] = joining_player
+            match_queues[desired_match_type]["waiting_json_data"] = json_data
+            match_queues[desired_match_type]["waiting_version"] = player_join_version
+            success = true
+        } else {
+            var starting_timer = this.match_types[desired_match_type][starting_timer]
+            var enforce_timer = this.match_types[desired_match_type][enforce_timer]
+            var minimum_time_per_choice = this.match_types[desired_match_type][minimum_time_per_choice]
+            waiting_player = match_queues[waiting_player]
+            waiting_json_data = match_queues[waiting_json_data]
+            waiting_version = waiting_json_data.version
+
+            if (waiting_version < player_join_version) {
+                // The player joining has a larger version,
+                // kick the player waiting from matchmaking, then
+                // put the joining player in their place.
+                send_join_version_error(waiting_player.ws)
+                this.remove_from_matchmaking(waiting_player)
+                match_queues[desired_match_type]["waiting_player"] = joining_player
+                match_queues[desired_match_type]["waiting_json_data"] = json_data
+                match_queues[desired_match_type]["waiting_version"] = player_join_version
+                success = true
+            } else if (waiting_version > player_join_version) {
+                // Lower version than room, probably need to update.
+                // Send error message.
+                send_join_version_error(ws)
+                return true
+            } else {
+                // Make a room and join it successfully.
+                const room = this.room_manager.create_room(waiting_player, 
+                    waiting_json_data, 
+                    starting_timer, 
+                    enforce_timer, 
+                    minimum_time_per_choice)
+                success = room.join(player)
+                awaiting_match_room = null
+            }
+        }
+
+        if (!success) {
```

Comment:
i think in the success case, all other users have to be informed about the queue states somehow.
Probably through the players_update message in server somehow, which should probably grab all queue states from Matchmaker or something.